polling/assets.go

```diff
@@ -298,6 +303,65 @@ func (d *DataSources) PullAllPEGAssets(oprversion uint8) (pa PegAssets, err erro
 	return pa, nil
 }
 
+// PullBestPrice pulls the best asset price we can find for a given asset.
+// Params:
+//		asset		Asset to pull pricing data
+//		reference	Time reference to determine 'staleness' from
+//		sources		Map of datasources to pull the price quote from.
+func (d *DataSources) PullBestPrice(asset string, reference time.Time, sources map[string]IDataSource) (pa PegItem, err error) {
+	if sources == nil {
+		// If our data sources passed in are nil, then we don't need to do cache wrapping.
+		// We should always have sources passed in, aside from unit tests.
+		sources = d.DataSources
+	}
+
+	// All the given data sources for the asset
+	sourceList := d.AssetSources[asset]
+
+	var prices []PegItem
+
+	// Eval all datasources from the reference time
+	for _, source := range sourceList {
+		var price PegItem
+		price, err = sources[source].FetchPegPrice(asset)
+		if err != nil {
+			continue
+		}
+
+		if price.Value != 0 {
+			prices = append(prices, price)
+
+			// We can break out if this is a 'good' price
+			//	Is it stale AND the market is open? We can expect stale quotes in a closed market
+			if reference.Sub(price.When) > d.staleDuration && IsMarketOpen(asset, reference) {
+				// This price quote is stale, keep fetching prices
+				continue
+			}
+
+			// This price is acceptable
+			break
+		}
+	}
+
+	// No prices found, and the last data source had an error
+	if err != nil && len(prices) == 0 {
```

Comment:
Currently, the polling can return a 0 price. If you, for example, had 0 data sources, the polling would not fail. There would be no error, and the list length is 0.

The opr creation would fail because a 0 price is an invalid opr. I do not want to enforce a non-zero in the polling, but rather do that in the opr validation. For example, `PEG` goes through this process, but returns a 0.

Reply:
I think a cleaner solution would be to have a non-configurable FixedPEG source that return 0, like FixedUSD, or just not calling a data source for peg/usd, rather than hiding the implicit value of PEG = 0 in a hard to understand if-statement

a quick solution would be to just put an explicit `if asset == "PEG" { return }` before the loop

Reply:
would this also just return a 0 if someone turned off all apis for a specific asset in the config?

Reply:
Looking back, I should add FixedPEG, but that would require a config update. I'll create a fixed PEG, and add it to the new config, and at some point we can enforce it going forward.

I will add an `if asset == "PEG" { return }` for now, and make an error if the other sources are not configured. At some point in the future, we can drop that and people will have to turn on the `FixedPEG` like they do the `FixedUSD`